File: CS/E3597/UnboundMasterDetailTree.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;

namespace treelist
{
    class UnboundMasterDetailTree
    {
        DataTable _dataSource;
        public DataTable DataSource
        {
            get { return _dataSource; }
            set
            {
                if (_dataSource == value)
                    return;
                _dataSource = value;
                //_masterTreeList.RefreshDataSource();
            }
        }
        TreeList _masterTreeList;
        TreeList MasterTreeList
        {
            get { return _masterTreeList; }
            set
            {
                if (_masterTreeList == value)
                    return;
                Detach();
                Attach(value);
            }
        }

        public UnboundMasterDetailTree(TreeList masterTreeList, DataTable dt)
        {
            MasterTreeList = masterTreeList;
            _masterTreeList.BeforeExpand += new BeforeExpandEventHandler(BeforeExpand);
            _masterTreeList.CellValueChanged += new CellValueChangedEventHandler(_masterTreeList_CellValueChanged);
            _dataSource = dt;
            foreach (DataRow row in _dataSource.Select())
            {
                TreeListNode node = _masterTreeList.AppendNode(new object[] { row["ID"], null, row["Value"] }, null, row);
                node.HasChildren = row.Table.ChildRelations.Count != 0;
            }
        }

        void _masterTreeList_CellValueChanged(object sender, CellValueChangedEventArgs e)
        {
            (e.Node.Tag as DataRow)[e.Column.FieldName] = e.Value;
        }

        void Detach()
        {
        }
        void Attach(TreeList treeList)
        {
            _masterTreeList = treeList;
        }
        private void BeforeExpand(object sender, BeforeExpandEventArgs e)
        {
            if (e.Node.Nodes.Count != 0) return;
            DataRow row = e.Node.Tag as DataRow;
            foreach (DataRow childRow in row.GetChildRows(row.Table.ChildRelations[0]))
            {
                TreeListNode node = MasterTreeList.AppendNode(new object[] { childRow["ID"], childRow["ParentID"], childRow["Value"] }, e.Node, childRow);
                node.HasChildren = childRow.Table.ChildRelations.Count != 0;
            }
        }


    }
}

File: VB/E3597/UnboundMasterDetailTree.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Data
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes

Namespace E3597
    Friend Class UnboundMasterDetailTree
        Private _dataSource As DataTable
        Public Property DataSource() As DataTable
            Get
                Return _dataSource
            End Get
            Set(ByVal value As DataTable)
                If _dataSource Is value Then
                    Return
                End If
                _dataSource = value
            End Set
        End Property
        Private _masterTreeList As treelist
        Private Property MasterTreeList() As treelist
            Get
                Return _masterTreeList
            End Get
            Set(ByVal value As treelist)
                If _masterTreeList Is value Then
                    Return
                End If
                Detach()
                Attach(value)
            End Set
        End Property

        Public Sub New(ByVal masterTreeList As treelist, ByVal dt As DataTable)
            Me.MasterTreeList = masterTreeList
            AddHandler _masterTreeList.BeforeExpand, AddressOf BeforeExpand
            AddHandler _masterTreeList.CellValueChanged, AddressOf _masterTreeList_CellValueChanged
            Dim rootNode As TreeListNode = Nothing

            _dataSource = dt
            For Each row As DataRow In _dataSource.Select()
                Dim node As TreeListNode = _masterTreeList.AppendNode(New Object() {row("ID"), Nothing, row("Value")}, rootNode, row)
                node.HasChildren = row.Table.ChildRelations.Count <> 0
            Next row
        End Sub

        Private Sub _masterTreeList_CellValueChanged(ByVal sender As Object, ByVal e As CellValueChangedEventArgs)
            TryCast(e.Node.Tag, DataRow)(e.Column.FieldName) = e.Value
        End Sub

        Private Sub Detach()
        End Sub
        Private Sub Attach(ByVal treeList As treelist)
            _masterTreeList = treeList
        End Sub
        Private Sub BeforeExpand(ByVal sender As Object, ByVal e As BeforeExpandEventArgs)
            If e.Node.Nodes.Count <> 0 Then
                Return
            End If
            Dim row As DataRow = TryCast(e.Node.Tag, DataRow)
            For Each childRow As DataRow In row.GetChildRows(row.Table.ChildRelations(0))
                Dim node As TreeListNode = MasterTreeList.AppendNode(New Object() {childRow("ID"), childRow("ParentID"), childRow("Value")}, e.Node, childRow)
                node.HasChildren = childRow.Table.ChildRelations.Count <> 0
            Next childRow
        End Sub


    End Class
End Namespace
